prog/prio.go

```diff
@@ -181,14 +181,21 @@ func normalizePrio(prios [][]int32) {
 
 // ChooseTable allows to do a weighted choice of a syscall for a given syscall
 // based on call-to-call priorities and a set of enabled and generatable syscalls.
-type ChoiceTable struct {
+type ChoiceTable interface {
```

Comment:
I am not sure this is really necessary. Is there anything the prog package that holds a pointer to ChoiceTable for prolonged duration? Functions like Generate/Mutate are fast, and it does not make sense to substitute the ChoiceTable while they are running.

I assume the problem is that we had some long running job (smash?) that persistent a pointer to old ChoiceTable. They should just not do that. If we have an atomic.Value pointing to the current *ChoiceTable, jobs should read it before each prog.Mutate/Generate call and pass the current value w/o persisting the pointer for prolonged time.

Theoretically with atomic.Values it's still possible to have more than 1 choice tables alive at the same time, if we add inputs very quickly during initial triage. It's possible to ensure we have at most 2 alive at the same time with https://pkg.go.dev/runtime#SetFinalizer, but I am not sure we want to do it. When we move fuzzer to syz-manager, it will be less of a problem.

Reply:
The problem was that now that we don't queue, but start jobs right away, they do actually start and only block on their first attempts to execute. And there are a lot of such blocked jobs. Even though mutation is fast, every goroutine continued to keep a reference to some older version of ChoiceTable.

Though now I wonder whether the problem was because we explicitly kept such a reference:

```
	choiceTable := fuzzer.ChoiceTable()
	for i := 0; i < iters; i++ {
		p := job.p.Clone()
		p.Mutate(rnd, prog.RecommendedCalls,
			choiceTable,
```

I'll try to roll it back and only scope `fuzzer.ChoiceTable()` to a `p.Mutate` argument. But I'm still worried that a one line change here (particularly one that seems reasonable) could increase memory consumptions many times. Too brittle.

Reply:
Avoiding a local reference to `prog.ChoiceTable` seems to have helped.